### ISSUE TITLE
The various OBJ calls are not thread safe

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#15674]
The OBJ database can be updated dynamically by user code using calls such as `OBJ_create` or `OBJ_add_sigid`. Unfortunately there are no locks at all used and therefore calling these functions is not thread safe.

**Comment by paulidale** [comment]
I'm not convinced that doing this for 1.1.1 is a wise move.

**Comment by t8m** [comment]
> I'm not convinced that doing this for 1.1.1 is a wise move.

I agree. I've never seen bug reports that would have thread unsafety of OBJ_ calls as the cause. IMO we could live with it even for 3.0 if it wasn't for the dynamic load/unload of providers and the callbacks added in #15681
